src/deadline/job_attachments/incremental_downloads/asset_downloader_with_manifest_aggregation.py

```diff
@@ -0,0 +1,145 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+
+from deadline.client.api._session import _get_queue_user_boto3_session
+from typing import List, Optional, Dict, Callable
+import boto3
+from ...job_attachments.download import (
+    get_output_manifests_by_asset_root,
+    merge_asset_manifests,
+)
+from ...job_attachments.api.attachment import _attachment_download_with_root_manifests
+from ...job_attachments.models import JobAttachmentS3Settings, FileConflictResolution
+from ...job_attachments.asset_manifests.base_manifest import BaseAssetManifest
+from ...job_attachments.incremental_downloads.session_action_processor import (
+    SessionAction,
+)
+
+
+class PartialDownloadException(Exception):
+    def __init__(self, message, failed_session_action_ids=None):
+        super().__init__(message)
+        self.failed_session_action_ids = (
+            failed_session_action_ids  # Store the list of failed IDs as an attribute.
+        )
+
+    def __str__(self):
+        return f"{self.args[0]} Failed Session action IDs: {self.failed_session_action_ids}"
+
+
+def aggregate_manifest_and_download_outputs(
+        boto3_session: boto3.Session,
+        session_actions: List[SessionAction],
+        farm_id: str,
+        queue_id: str,
+        file_conflict_resolution: FileConflictResolution,
+        path_mapping_rules: Optional[str] = None,
+        print_function_callback: Callable[[str], None] = lambda msg: None,
+) -> None:
+    """
+    Aggregate manifests and download outputs for the given session actions.
```

Comment:
This is only for taskRun session actions, correct? Would be good to specify that.